sdk/daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/ValueTranslatorSpec.scala

```diff
@@ -429,29 +496,65 @@ class ValueTranslatorSpec
             typ shouldBe TRecordUpgradable
           },
         ),
+        (
+          TRecordUpgradable,
+          ValueRecord(
+            "",
+            ImmArray(), // missing a non-optional field
+          ),
+          { case Error.Preprocessing.TypeMismatch(typ, _, _) =>
+            typ shouldBe TRecordUpgradable
+          },
+        ),
         (
           TRecordUpgradable,
           ValueRecord(
             "",
             ImmArray(
               "fieldA" -> aInt,
-              "fieldB" -> someParty, // Here the field has type Party instead of Text
-              "fieldC" -> none,
+              "fieldB" -> someText,
+              "fieldC" -> someParty,
+              "fieldD" -> aInt, // extra non-optional field
             ),
           ),
-          { case Error.Preprocessing.TypeMismatch(typ, value, _) =>
-            typ shouldBe t"Text"
-            value shouldBe aParty
+          { case Error.Preprocessing.TypeMismatch(typ, _, _) =>
+            typ shouldBe TRecordUpgradable
           },
         ),
         (
-          t"Mod:Variant Int64 Text",
+          TVariantUpgradable,
           ValueVariant("", "ConsB", aInt), // Here the variant has type Text instead of Int64
           { case Error.Preprocessing.TypeMismatch(typ, value, _) =>
             typ shouldBe t"Text"
             value shouldBe aInt
           },
         ),
+        (
+          TVariantUpgradable,
+          ValueVariant("", "ConsC", aInt), // ConsC is not a constructor of Mod:Variant
+          {
+            case Error.Preprocessing.Lookup(
+                  LookupError.NotFound(
+                    Reference.DataVariantConstructor(_, consName),
+                    Reference.DataVariantConstructor(_, _),
+                  )
+                ) =>
+              consName shouldBe "ConsC"
+          },
+        ),
+        (
+          TEnumUpgradable,
+          ValueEnum("", "Cons3"), // Cons3 is not a constructor of Mod:Enum
+          {
+            case Error.Preprocessing.Lookup(
+                  LookupError.NotFound(
+                    Reference.DataEnumConstructor(_, consName),
+                    _,
```

Comment:
Can we assert that this wildcard has the shape `Reference.DataEnumConstructor(_, _)`?